java-checks/src/main/java/org/sonar/java/checks/AssertionFailInCatchBlockCheck.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.Lists;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.NameCriteria;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S3658")
+public class AssertionFailInCatchBlockCheck extends AbstractMethodDetection {
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return Lists.newArrayList(
+      MethodMatcher.create().typeDefinition("org.junit.Assert").name("fail").withNoParameterConstraint(),
+      MethodMatcher.create().typeDefinition("junit.framework.Assert").name(NameCriteria.startsWith("fail")).withNoParameterConstraint(),
+      MethodMatcher.create().typeDefinition("org.fest.assertions.Fail").name(NameCriteria.startsWith("fail")).withNoParameterConstraint());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (isInCatch(mit)) {
+      reportIssue(mit, "Remove this failure assertion and simply add the exception type to the method signature.");
+    }
+  }
+
+  private static boolean isInCatch(Tree tree) {
+    if (tree.parent() == null) {
```

Comment:
> This whole method can probably be refactored like : 

Many thanks, I'm always a newbie in tree API usage :)

> It can also probably be improved by stopping the recursion when one of the parent is a method, no need to go up to the compilation unit.

Initially I left all the tree to catch this corner case (but without unit test) :
```
public class MyTest {

  static {
    try {
      File.createTempFile(null, null);
    } catch (IOException e) {
      fail(e.getMessage()); // Noncompliant {{...}}
      // Should be : throw new UnsupportedOperationException(e);
    }
  }
```
But the current code doesn't catch it => question : This use case (`fail` usage in `static` block of class) is a bad practice or not ? 
(I yes, I need to investiguate why `MethodMatcher` doesn't detect this case ...)